mcs/class/corlib/System/Array.cs

```diff
@@ -583,7 +583,7 @@ public static void Copy (Array sourceArray, int sourceIndex, Array destinationAr
 			if (destinationIndex < 0)
 				throw new ArgumentOutOfRangeException ("destinationIndex", "Value has to be >= 0.");
 
-			if (FastCopy (sourceArray, sourceIndex, destinationArray, destinationIndex, length))
+			if (FastCopy (ref sourceArray, sourceIndex, ref destinationArray, destinationIndex, length))
```

Comment:
I wonder if we need extra locals here.

If we had a caller like `Copy (o.member1, 0, o.member2, 0, len)` and then `Copy` is inlined, won't we potentially end up with passing a heap address to `FastCopy` here? 

Maybe we should have `FastCopy` that assigns the params to locals and have it call a private `FastCopy_icall` with ref arguments.

Reply:
Oh good point. Mark it as noinline?

Reply:
But does it matter? If only read? Maybe start adding const to “ref ref”?

Reply:
suppose we have
```
  class C {
     object member;
   }
  C c = new C ();
  c.member = new object ();
```

if we have `ref c.member` that's a pointer into the middle of `c`.  So we'll pin `c`.  But nothing is pinning the object that `c.member` currently points to. So the GC could move it.  so if on the native side you have `MonoObject * volatile * ptr` and if you're not careful about always using `*ptr` (instead of a local initialized with the present value of `*ptr`) the GC could move the object and you will be pointing at garbage.